src/sentry/roles/manager.py

```diff
@@ -31,55 +51,155 @@ def has_scope(self, scope: str) -> bool:
 
-        self._top_dog = role_list[-1]
+    def can_manage(self, other: TeamRole) -> bool:
+        return self.priority >= other.priority
 
-    def __iter__(self) -> Iterable[Role]:
-        yield from self._roles.values()
+
+class RoleLevel(Generic[R]):
+    """Represent the set of all roles at one level (org or team)."""
+
+    def __init__(self, roles: Iterable[R], default_id: str | None = None) -> None:
+        self._priority_seq = tuple(sorted(roles, key=lambda r: r.priority))
+        self._id_map = OrderedDict((r.id, r) for r in self._priority_seq)
+
+        self._choices = tuple((r.id, r.name) for r in self._priority_seq)
+        self._default = self._id_map[default_id] if default_id else self._priority_seq[0]
+        self._top_dog = self._priority_seq[-1]
+
+    def __iter__(self) -> Iterable[R]:
+        yield from self._priority_seq
 
     def can_manage(self, role: str, other: str) -> bool:
         return self.get(role).priority >= self.get(other).priority
 
-    def get(self, id: str) -> Role:
-        return self._roles[id]
+    def get(self, id: str) -> R:
+        return self._id_map[id]
 
-    def get_all(self) -> Iterable[Role]:
-        return list(self._roles.values())
+    def get_all(self) -> Sequence[R]:
+        return self._priority_seq
 
-    def get_choices(self) -> Iterable[Tuple[str, str]]:
+    def get_choices(self) -> Sequence[Tuple[str, str]]:
         return self._choices
 
-    def get_default(self) -> Role:
+    def get_default(self) -> R:
         return self._default
 
-    def get_top_dog(self) -> Role:
+    def get_top_dog(self) -> R:
         return self._top_dog
 
-    def with_scope(self, scope: str) -> Iterable[Role]:
+    def with_scope(self, scope: str) -> Iterable[R]:
         for role in self.get_all():
             if role.has_scope(scope):
                 yield role
 
-    def with_any_scope(self, scopes: Iterable[str]) -> Iterable[Role]:
+    def with_any_scope(self, scopes: Iterable[str]) -> Iterable[R]:
         for role in self.get_all():
             if any(role.has_scope(scope) for scope in scopes):
                 yield role
+
+
+class RoleManager:
+    def __init__(
+        self,
+        org_config: Iterable[Mapping[str, str]],
+        team_config: Iterable[Mapping[str, str]],
+        default_org_role: str | None = None,
+    ) -> None:
+        self.organization_roles: RoleLevel[OrganizationRole] = RoleLevel(
+            (
+                OrganizationRole.from_config(self, idx, **role_cfg)
+                for idx, role_cfg in enumerate(org_config)
+            ),
+            default_org_role,
+        )
+
+        self.team_roles: RoleLevel[TeamRole] = RoleLevel(
+            TeamRole.from_config(self, idx, **role_cfg) for idx, role_cfg in enumerate(team_config)
+        )
+
+        self._minimum_team_role_map = self._make_minimum_team_role_map(
+            self.organization_roles, self.team_roles
+        )
+
+    @staticmethod
+    def _make_minimum_team_role_map(
```

Comment:
This is a map of minimum team roles to orgs?
so in most cases (assuming most ppl only have 1 org) this would be a single org with a single minimum team role?

Reply:
No, it's a map from organization roles to the corresponding minimum team roles. There's a single global map derived from what's in the server config.